numpyro/distributions/util.py

```diff
@@ -13,21 +13,144 @@
+                    1. / (12 * (k + 1)) - 1./(360 * (k + 1) ** 3) + 1./(1260 * (k + 1) ** 5))
+
+
+def _binomial_btrs(key, p, n):
+    """
+    Based on the transformed rejection sampling algorithm (BTRS) from the
+    following reference:
+
+    Hormann, "The Generation of Binonmial Random Variates"
+    (https://core.ac.uk/download/pdf/11007254.pdf)
+    """
+    def _btrs_body_fn(val):
+        _, key, p, n, tr_params, _, _ = val
+        key, key_u, key_v = random.split(key, 3)
+        u = random.uniform(key_u)
+        v = random.uniform(key_v)
+        u = u - 0.5
+        k = np.floor((2 * tr_params.a / (0.5 - np.abs(u)) + tr_params.b) * u + tr_params.c).astype(n.dtype)
+        return k, key, p, n, tr_params, u, v
+
+    def _btrs_cond_fn(val):
+        def accept_fn(k, p, n, tr_params, u, v):
+            m = np.floor((n + 1) * p).astype(n.dtype)
+            q = 1 - p
+            log_f = (m + 0.5) * np.log((m + 1.) / (n - m + 1.) * (q / p)) + \
+                    (n + 1.) * np.log((n - m + 1.) / (n - k + 1.)) + \
+                    (k + 0.5) * np.log((n - k + 1.) / (k + 1.) * p / q) + \
+                    (stirling_approx_tail(m) + stirling_approx_tail(n - m)) - \
+                    (stirling_approx_tail(k) - stirling_approx_tail(n - k))
+            log_g = (tr_params.a / (0.5 - np.abs(u)) ** 2) + tr_params.b
+            return np.log(v) <= log_f + log_g - np.log(tr_params.alpha)
+
+        k, key, p, n, tr_params, u, v = val
+        early_accept = (np.abs(u) <= tr_params.u_r) & (v <= tr_params.v_r)
+        return lax.cond(early_accept,
+                        (),
+                        lambda _: False,
+                        (k, p, n, tr_params, u, v),
+                        lambda v: lax.cond((k < 0) | (k > n),
+                                           (),
+                                           lambda _: True,
+                                           v,
+                                           lambda v: ~accept_fn(*v)))
+
+    tr_params = _get_tr_params(n, p)
+    ret = lax.while_loop(_btrs_cond_fn, _btrs_body_fn,
+                         (n + np.ones_like(n), key, p, n, tr_params, np.ones_like(p), np.ones_like(p)))
+    return ret[0]
+
+
+def _binomial_inversion(key, p, n):
+    def _binom_inv_body_fn(val):
+        i, key, p, n, geom_acc = val
+        key, key_u = random.split(key)
+        u = random.uniform(key_u)
+        geom = np.clip(np.ceil(np.log(1 - u) / np.log(1 - p)), a_min=1., a_max=n + 1.)
+        geom_acc += geom
+        return i + 1, key, p, n, geom_acc
+
+    def _binom_inv_cond_fn(val):
+        i, _, _, n, geom_acc = val
+        return geom_acc <= n
+
+    ret = lax.while_loop(_binom_inv_cond_fn, _binom_inv_body_fn,
+                         (np.zeros_like(n), key, p, n, np.zeros_like(p)))
+    return ret[0] - 1
+
+
+def _binomial_dispatch(key, p, n):
+    def dispatch(key, p, n):
+        eps = 1e-6
+        p = np.clip(p, a_min=eps, a_max=1 - eps)
```

Comment:
This is only for initial jit tracing, For p=0., 1., I found that we were getting stuck during the jit tracing phase itself. I'll check if we still need this.

Reply:
It seems to not necessary because of the guard at `lax.cond`. I think JAX is smarter now so that it won't run through the whole while_loop (only 1 step is required I guess) in the `false` branch during tracing. But that's for some toy examples that I played with. I might miss something here.